### DISCUSSION TITLE
exchangelib.errors.ErrorServerBusy

### DISCUSSION THREAD

**Discussion by Tre-Seibert** [ecederstrand/exchangelib#1225] [Q&A] [answered]
I understand that the exchange server is under stress due to the data size of the api call. I believe I am pulling all of the calendar items, however I just need the latest 3 calendar items.  I just want to know if there is anything I can do to avoid this error besides using fault tolerance, as that also doesn't always work. Is there a way to only grab the latest 3 calendar items in order to reduce the data size?

```python
      calendar_items = account.calendar.all()[::-1][-3:]
        for item in reversed(list(calendar_items)):
            if isinstance(item, CalendarItem):
                event_data = {
                    'subject': item.subject,
                    'start': item.start.astimezone(timezone("US/Eastern")).strftime('%m/%d/%Y %I:%M %p'),
                    'end': item.end.astimezone(timezone("US/Eastern")).strftime('%m/%d/%Y %I:%M %p'),
                    'location': item.location,
                    'body': item.body
                }
                calendar_events.append(event_data)
```

**Comment by ecederstrand** [comment]
Can you expand on where you see fault tolerance not always working?

To meaningfully get "latest 3 calendar items" you would need to sort on a field, since items are not guaranteed to be returned in any specific order by default. When you have done that, you can use slicing to limit the number of items fetched. Additionally, you can limit the fields you want to fetch. If you want to make sure only calendar items are returned, you can filter on `item_class`:

In summary, here's what you can do:

```python
last_three = account.calendar\
                 .filter(item_class="IPF.Appointment")\
                 .only("subject", "start", "end", "location", "body")\
                 .order_by("-datetime_received")[:3]
```

**Reply by Tre-Seibert** [reply]
Fault tolerance doesn't always work because some users on my 365 have a large amount of Calendar Items. Even with fault tolerance it would take upwards to an hour to complete the request, also this error occurs a lot quicker than it takes to load all of some of my users calendar items. Instead of retrieving all of them in the api call I would like to just get the 3 latest ones, based on the latest end time of a calendar item.

**Reply by ecederstrand** [reply]
The fault tolerance settings apply to a single HTTP request. That should never take an hour to return. When exchangelib fetches large numbers of items, it uses pagination to do so, so only e.g. 100 items are returned in a single request.

If you have cases where fault tolerance didn't work for you where you expected it to, then please open an issue in the isue tracker describing the issue and adding stack traces or error messages you got.

**Comment by Tre-Seibert** [comment, marked as answer]
Solved by using the following:

```python
        for item in account.calendar.all().only("subject", "start", "end", "location", "body"):
            calendar_items.append(item)
            print(calendar_items)
            if len(calendar_items) == 3:
                break
```

**Reply by ecederstrand** [reply]
If you want to just get the first three itrems, then the recommended way to do it in exchangelib is by using slicing:

```python
for item in account.calendar.all().only("subject", "start", "end", "location", "body")[:3]:
    print(item)
```

That will fetch exactly three items from the server.

**Reply by Tre-Seibert** [reply]
Thank you for the clarification!
